Refactor models to keep profiles (software, compute, network, dbParam…

tilaksatra opened the pull request:
Following are the code changes to refactor models to keep profiles (software, compute, network, etc) as optional and use default if not specified.

Changes / Intention (or, Motivation) behind the said changes:

- Implemented an Eagerly Fetched and Facade Design Pattern based approach to fill custom profiles
  - By default, the code will redirect to getOOBProfiles.
  - enrichProfilesMap() function is defined which based on the custom profiles passed in YAML does the following:
     (a) Checks whether profiles are really provided or else proceed with OOB
     (b) If provided, checks which specific profile is provided through YAML, and performs checks to validate from the available profile
     (c) Performs Eagerly Fetched Error Handling by throwing error as soon as any of the custom profile is not found to be matched from the available list of profiles each for Software, Compute, Network and DBParam Profiles


Testing:

Following Scenarios were tested for Postgres Database Type:

(a) Custom Compute Small Profile was provided and database provision request was sent successfully to the NDB
(b) Custom Compute Small Profile and Incorrect "ID" for HBA Software Profile was provided 
     Result: Database Provision Request Failed as the Custom Software Profile was incorrect, logged error message and thus did not perform database provisioning.
(c) Custom Compute Small Profile and HBA Software Profile was provided 
     Result: Database Provision Request Succeeded as the Custom Software Profile was correct, logged proper messages and thus perform database provisioning.     
     

NOTE: All requests were sent to NDB; however, on NDB provisioning failed with the below error message:

Error Message:
Failed to create DBServer VM. Reason : "'NoHostResources: No host has enough available memory for VM ntnx3_db_version_1.1_VM (35daa134-4fca-4957-b9e4-d604a1a7e693) that uses 2 vCPUs and 4096MB of memory. You could try downsizing the VM, increasing host memory, power off some VMs, or moving the VM to a different host. Maximum allowable VM size is approximately 2521 MB'"

manavrajvanshi commented on api/v1alpha1/ndb_api_helpers.go:
```diff
@@ -153,6 +162,115 @@ func GenerateProvisioningRequest(ctx context.Context, ndbclient *ndbclient.NDBCl
+	isValidProfile = false
+	err = nil
+	matchedProfile = ProfileResponse{}
+	for _, eachProfile := range profiles {
+		if eachProfile.Type == profileType {
+			if (eachProfile.Id == customProfile.Id) && (eachProfile.LatestVersionId == customProfile.VersionId) {
+				log.Info("Profile Matched for custom profile", "profileType", profileType, "profileValue", eachProfile)
+				isValidProfile = true
+				matchedProfile = eachProfile
+				break
+			}
+		}
+	}
```
Can we reuse the util/filter function here?

tilaksatra replied:
This filter funtion has been removed and replaced by the filtering available through util package

manavrajvanshi commented on config/samples/ndb_v1alpha1_database.yaml:
```diff
@@ -4,17 +4,24 @@ metadata:
-    server: https://[NDB IP]:8443/era/v0.9
+    clusterId: "0f812c8e-0607-46cf-886e-706ce7d30b6a"
+    credentialSecret : ntnx3-ndb-secret
+    server: https://eraybgyd9x86x-era.nutanixtestdrive.com/era/v0.9
     skipCertificateVerification: true
   databaseInstance:
-    databaseInstanceName: "Database Instance Name"
+    databaseInstanceName: "ntnx3_db_version_1.1"
     databaseNames:
       - database_one
       - database_two
       - database_three
-    credentialSecret: db-instance-secret-name
+    credentialSecret: ntnx3-db-secret
     size: 10
     timezone: "UTC"
     type: postgres
+    profiles:
+      compute:
+        id: "f6594223-dc7d-45da-ab40-07ce8c2dcdac"
+        versionId: "23fffd73-f582-4cd3-82c0-db09e580a937"
+      software:
+        id: "780db288-93eb-4d59-b90d-e34e47adb3df"
+        versionId: "97ae4340-37f3-403b-bd41-4f11965ea6ef"
```
Please don't change the default values in the sample yaml files. We can add the fields for profiles, but shouldn't populate with some specific profile id and version values.

tilaksatra replied:
Yes, I'll be reverting these check ins

manavrajvanshi commented on config/samples/secret.yaml:
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ntnx3-ndb-secret
+type: Opaque
+stringData:
+  username: ntnx3
+  password: ntnx3
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ntnx3-db-secret
+type: Opaque
+stringData:
+  password: NTNX3
+  ssh_public_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQCslQJ1257QMu3TES1R3m8vYkRHGywcuCBq1c3RRMbAwe7nNbmWpcMwEoXhCXCqxxo12XCM59svv7XvoLPnCAUMxQap1BjEbRwP4BbFIbK+OLzJ+22j5FLj4fAGJx62s6BvG1+QeRCoW96C+qR92FP6f8wNmLfiskealYR5UsXaM8HOdOhCkVvm4C1A9L0V+Vr257TvW84mR9rxNws4xqSevkS51yq5IW4h+JmbkWLnFcbnGhaywKUOrS0X8wwv+XtxIslf+OQk1PAcvX5tsQ1b70qcrpp+ZxtmMHTOZdsnUqeFNb7AYa7PoR7hHbLxcBZwqXA61OeLsydScqHQw9oig ...
+
```
Please remove this file from your check-in.

tilaksatra replied:
Yes, I'll be reverting these check ins

manavrajvanshi commented on ndbclient/ndbclient.go:
```diff
@@ -48,6 +47,7 @@ func NewNDBClient(username, password, url, caCert string, skipVerify bool) *NDBC
 func (ndbClient *NDBClient) Get(path string) (*http.Response, error) {
 	url := ndbClient.url + "/" + path
 	req, err := http.NewRequest("GET", url, nil)
+	req.Header.Add("Cookie", "eraAuth=eyJhbGciOiJSUzUxMiJ9")
 	if err != nil {
 		// fmt.Println(err)
 		return nil, err
```
Please don't check-in the header changes, these are just to circumvent the test-drive errors.

tilaksatra replied:
Yes, I'll be reverting these check ins

manavrajvanshi commented on api/v1alpha1/ndb_api_helpers.go:
```diff
@@ -153,6 +162,115 @@ func GenerateProvisioningRequest(ctx context.Context, ndbclient *ndbclient.NDBCl
+	isValidProfile = false
+	err = nil
+	matchedProfile = ProfileResponse{}
+	for _, eachProfile := range profiles {
+		if eachProfile.Type == profileType {
+			if (eachProfile.Id == customProfile.Id) && (eachProfile.LatestVersionId == customProfile.VersionId) {
```
Also, as discussed earlier.. can we add a method to include profile filtering based on profile names as well?

tilaksatra replied:
This can be handled in later iterations. Currently, I'm majorly focusing on ID & VersionID
Once, this is tested, extending to name should be quite easy. That's my thinking

manavrajvanshi reviewed (changes requested):
Some comments around the DRY principle and wrongly updated yaml files.
Please add Unit Tests for the changes you've made and make sure the existing tests are passing before you push the changes.

tilaksatra commented on api/v1alpha1/ndb_api_helpers.go:
```diff
@@ -50,6 +50,14 @@ func GenerateProvisioningRequest(ctx context.Context, ndbclient *ndbclient.NDBCl
 
 	database_names := strings.Join(dbSpec.Instance.DatabaseNames, ",")
 
+	profiles, err := GetAllProfiles(ctx, ndbclient)
+
+	err = enrichProfilesMap(ctx, dbSpec, profiles, profilesMap, dbSpec.Instance.Type)
+	if err != nil {
+		log.Error(err, "Error occurred while performing enrichment of custom profiles", "database name", dbSpec.Instance.DatabaseInstanceName, "database type", dbSpec.Instance.Type)
+		return
+	}
+
```
Yes, call has been moved to OOB()

tilaksatra commented on api/v1alpha1/ndb_api_helpers.go:
```diff
@@ -153,6 +162,115 @@ func GenerateProvisioningRequest(ctx context.Context, ndbclient *ndbclient.NDBCl
+			log.Info("No enrichment for software profiles as no custom profile received for Software. Hence, proceeding with default OOB software profile")
+		} else {
+			isValidProfile, matchedProfile, errorThroughChecks := performProfileAvailabilityCheck(ctx, dbEngineSpecificProfiles, softwareProfile, PROFILE_TYPE_SOFTWARE)
+			if errorThroughChecks != nil {
+				//log.Error(err, "")
+				return errorThroughChecks
+			}
+			if isValidProfile {
+				profilesMap[PROFILE_TYPE_SOFTWARE] = matchedProfile
+			}
+		}
+
+		//Custom Compute Profile Check and overriding the default values
+		computeProfile := dbSpec.Instance.Profiles.Compute
+		if computeProfile == (Profile{}) {
+			log.Info("No enrichment for compute profiles as no custom profile received for Compute. Hence, proceeding with default OOB compute profile")
+		} else {
+			isValidProfile, matchedProfile, errorThroughChecks := performProfileAvailabilityCheck(ctx, genericProfiles, computeProfile, PROFILE_TYPE_COMPUTE)
+			if errorThroughChecks != nil {
+				//log.Error(err, "")
+				return errorThroughChecks
+			}
+			if isValidProfile {
+				profilesMap[PROFILE_TYPE_COMPUTE] = matchedProfile
+			}
+		}
+
+		//Custom Network Profile Check and overriding the default values
+		networkProfile := dbSpec.Instance.Profiles.Network
+		if networkProfile == (Profile{}) {
+			log.Info("No enrichment for network profiles as no custom profile received for Network. Hence, proceeding with default OOB network profile")
+		} else {
+			isValidProfile, matchedProfile, errorThroughChecks := performProfileAvailabilityCheck(ctx, dbEngineSpecificProfiles, networkProfile, PROFILE_TYPE_NETWORK)
+			if errorThroughChecks != nil {
+				//log.Error(err, "")
+				return errorThroughChecks
+			}
+			if isValidProfile {
+				profilesMap[PROFILE_TYPE_NETWORK] = matchedProfile
+			}
```
Yes, I've created a profile loop, modularized the code and that should solve the concern

tilaksatra commented on api/v1alpha1/ndb_api_helpers.go:
```diff
@@ -153,6 +162,115 @@ func GenerateProvisioningRequest(ctx context.Context, ndbclient *ndbclient.NDBCl
+
+func performProfileAvailabilityCheck(ctx context.Context, profiles []ProfileResponse, customProfile Profile, profileType string) (isValidProfile bool, matchedProfile ProfileResponse, err error) {
+	log := ctrllog.FromContext(ctx)
+	isValidProfile = false
+	err = nil
```
Yes, these are redundant, have updated in the recent commits

tilaksatra commented:
Latest commits - Incorporated Review Comments:-

1 Maximized the usage of the DRY principle and eliminated redundant code
2 Added Unit Tests for the new code added and fixed the existing test cases w.r.t logging messages, function names, and their execution failures due to new function definition
3 Renamed GetOOBProfiles function names and introduced new modular functions where applicable
4 Topology being considered in the profile filtering/matching criteria
5 Defaulted to the usage of using util.filter function for profile matching instead of the creation of a new matching function
6 Removed modified config files from check-in

NOTE:-
7 Custom Profile Based on Names need to be worked upon after a call to clear certain questions that we have!

manavrajvanshi commented:
Closing since this is duplicate of #87.
